src/engine/strat_engine/thinpool/thinpool.rs

```diff
@@ -277,16 +284,20 @@ impl ThinPool {
                     }
                 }
 
-                let usage = &status.usage;
-                if usage.used_meta > usage.total_meta - META_LOWATER {
-                    // TODO: Extend meta device
+                if status.usage.used_meta >
+                   cmp::max(status.usage.total_meta, META_LOWATER) - META_LOWATER {
+                    match self.extend_thinpool_meta(dm, status.usage.total_meta, bd_mgr) {
+                        #![allow(single_match)]
+                        Ok(_) => {}
+                        Err(_) => {} // TODO: Take pool offline?
+                    }
                 }
 
-                if usage.used_data > usage.total_data - DATA_LOWATER {
+                if status.usage.used_data > status.usage.total_data - DATA_LOWATER {
```

Comment:
Do the same trick as above here, to ensure no overflow on subtraction.

Reply:
ok - I broke both out into commits.